sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -552,13 +581,28 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
         val newLeft = removeRedundantAliases(left, excluded ++ right.outputSet)
         val newRight = removeRedundantAliases(right, excluded ++ newLeft.outputSet)
         val mapping = AttributeMap(
-          createAttributeMapping(left, newLeft) ++
-          createAttributeMapping(right, newRight))
-        val newCondition = condition.map(_.transform {
-          case a: Attribute => mapping.getOrElse(a, a)
-        })
+          createAttributeMapping(left, newLeft) ++ createAttributeMapping(right, newRight))
+        val newCondition = condition.map(mapAttribute(_, mapping))
         Join(newLeft, newRight, joinType, newCondition, hint)
 
+      case p: Project =>
+        val newChild = removeRedundantAliases(p.child, excluded)
+        val mapping = AttributeMap(createAttributeMapping(p.child, newChild))
+        Project(removeRedundantAliases(mapAttributes(p.projectList, mapping), excluded), newChild)
+
+      case a: Aggregate =>
+        val newChild = removeRedundantAliases(a.child, excluded)
+        val mapping = AttributeMap(createAttributeMapping(a.child, newChild))
+        Aggregate(mapAttributes(a.groupingExpressions, mapping),
+          removeRedundantAliases(mapAttributes(a.aggregateExpressions, mapping), excluded),
+          newChild)
+
+      case w: Window =>
```

Comment:
why do we need to deal with these 3 nodes separately now?

Reply:
The `removeRedundantAliases()` is dealing with the output defining named expression lists as a whole (`Project.projectList`, `Aggregate.aggregateExpressions`, ...) to be able to keep those aliases that are required to keep the output attributes unique.